src/librustc/middle/check_const.rs

```diff
@@ -251,9 +251,13 @@ impl<'a, 'tcx, 'v> Visitor<'v> for CheckCrateVisitor<'a, 'tcx> {
                 b: &'v ast::Block,
                 s: Span,
                 fn_id: ast::NodeId) {
-        assert!(self.mode == Mode::Var);
+        let (old_mode, old_qualif) = (self.mode, self.qualif);
+        self.mode = Mode::Var;
+        self.qualif = ConstQualif::empty();
```

Comment:
You don't have to *copy* `with_mode`, you can just call it (and replace `self` in the two original statements with `v`, or whatever you choose for the name of the closure argument).